packages/sdk/android/src/main/java/com/primerioreactnative/components/manager/ach/PrimerRNHeadlessUniversalCheckoutStripeAchUserDetailsComponent.kt

```diff
@@ -0,0 +1,310 @@
+      PrimerHeadlessUniversalCheckoutAchManager(currentViewModelStoreOwner)
+          .provide<StripeAchUserDetailsComponent>("STRIPE_ACH")
+    }
+        .onSuccess { component = it }
+        .onFailure { 
+          val exception = ErrorTypeRN.NativeBridgeFailed errorTo "$INITIALIZATION_ERROR: ${it.message}"
+          promise.reject(exception.errorId, exception.description)
+          return
+         }
+
+    val lifecycleScope = (activity as LifecycleOwner).lifecycleScope
+
+    job =
+        lifecycleScope.launch {
+          coroutineScope {
+            launch { configureStepListener() }
+
+            launch { configureValidationListener() }
+
+            launch { configureErrorListener() }
+            promise.resolve(null)
+          }
+        }
+  }
+
+  @ReactMethod
+  fun start(promise: Promise) {
+    if (component == null) {
+      val exception = ErrorTypeRN.NativeBridgeFailed errorTo UNINITIALIZED_ERROR
+      promise.reject(exception.errorId, exception.description)
+    } else {
+      component?.start()
+      promise.resolve(null)
+    }
+  }
+
+  @ReactMethod
+  fun submit(promise: Promise) {
+    if (component == null) {
+      val exception = ErrorTypeRN.NativeBridgeFailed errorTo UNINITIALIZED_ERROR
+      promise.reject(exception.errorId, exception.description)
+    } else {
+      component?.submit()
+      promise.resolve(null)
+    }
+  }
+
+  @ReactMethod
+  fun onSetFirstName(firstName: String, promise: Promise) {
+    if (component == null) {
+      val exception = ErrorTypeRN.NativeBridgeFailed errorTo UNINITIALIZED_ERROR
+      promise.reject(exception.errorId, exception.description)
+    } else {
+      component?.updateCollectedData(AchUserDetailsCollectableData.FirstName(firstName))
+      promise.resolve(null)
+    }
+  }
+
+  @ReactMethod
+  fun onSetLastName(lastName: String, promise: Promise) {
+    if (component == null) {
+      val exception = ErrorTypeRN.NativeBridgeFailed errorTo UNINITIALIZED_ERROR
+      promise.reject(exception.errorId, exception.description)
+    } else {
+      component?.updateCollectedData(AchUserDetailsCollectableData.LastName(lastName))
+      promise.resolve(null)
+    }
+  }
+
+  @ReactMethod
+  fun onSetEmailAddress(emailAddress: String, promise: Promise) {
+    if (component == null) {
+      val exception = ErrorTypeRN.NativeBridgeFailed errorTo UNINITIALIZED_ERROR
+      promise.reject(exception.errorId, exception.description)
+    } else {
+      component?.updateCollectedData(AchUserDetailsCollectableData.EmailAddress(emailAddress))
+      promise.resolve(null)
+    }
+  }
+
+  private suspend fun configureErrorListener() {
+    component?.componentError?.collectLatest { error ->
+      sendEvent(
+          PrimerHeadlessUniversalCheckoutComponentEvent.ON_ERROR.eventName,
+          JSONObject().apply {
+            put(
+                "errors",
+                JSONArray().apply { put(JSONObject(json.encodeToString(error.toPrimerErrorRN()))) }
```

Comment:
[nit]: we seem to have this in every component, maybe worth centralising this and use in all components?

Reply:
Will do.